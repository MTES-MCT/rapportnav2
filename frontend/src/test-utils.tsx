import React, { ReactElement } from 'react'
import { render, renderHook, RenderOptions } from '@testing-library/react'
import UIThemeWrapper from './ui/ui-theme-wrapper'
import { BrowserRouter } from 'react-router-dom'

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <UIThemeWrapper>
      <BrowserRouter>{children}</BrowserRouter>
    </UIThemeWrapper>
  )
}

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, 'wrapper'>) =>
  render(ui, { wrapper: AllTheProviders, ...options })

const customRenderHook = (hook: any, options?: Omit<RenderOptions, 'wrapper'>) =>
  renderHook(() => hook(), { wrapper: AllTheProviders, ...options })

export * from '@testing-library/react'
export { customRender as render }
export { customRenderHook as renderHook }
