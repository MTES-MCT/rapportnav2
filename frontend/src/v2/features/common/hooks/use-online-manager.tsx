import { useEffect, useState } from 'react'
import { onlineManager } from '@tanstack/react-query'
import { useOfflineSince } from './use-offline-since.tsx'

/**
 * Offline-first toggle hook.
 * - Detects physical network availability via browser events.
 * - Allows users to manually switch to offline mode even when network is available.
 * - Integrates with TanStack Query's onlineManager for query behavior.
 */
export function useOnlineManager() {
  const { offlineSince, setOfflineSince } = useOfflineSince()
  // Physical network availability
  const [hasNetwork, setHasNetwork] = useState(() => navigator.onLine)

  // Effective online state from TanStack Query
  const [isOnline, setIsOnline] = useState(() => onlineManager.isOnline())

  // Subscribe to TanStack Query onlineManager changes
  useEffect(() => {
    const handleChange = online => setIsOnline(online)
    const unsubscribe = onlineManager.subscribe(handleChange)
    return unsubscribe
  }, [])

  const computeStatus = (hasNetwork: boolean, manualOffline: boolean) => {
    const effective = hasNetwork && !manualOffline
    if (effective !== onlineManager.isOnline()) {
      onlineManager.setOnline(effective)
      setIsOnline(effective)

      setOfflineSince(effective)
    }
  }

  // Sync TanStack Query & internal state whenever network or manualOffline changes
  useEffect(() => {
    computeStatus(hasNetwork, !!offlineSince)
  }, [computeStatus, hasNetwork, offlineSince, setOfflineSince])

  // Listen to browser connectivity changes
  useEffect(() => {
    const onOnline = () => setHasNetwork(true)
    const onOffline = () => setHasNetwork(false)

    window.addEventListener('online', onOnline)
    window.addEventListener('offline', onOffline)

    return () => {
      window.removeEventListener('online', onOnline)
      window.removeEventListener('offline', onOffline)
    }
  }, [])

  // User toggle function
  const toggleOnline = (shouldBeOnline: boolean) => {
    if (shouldBeOnline) {
      // Only clear manualOffline if there's physical network
      if (hasNetwork) {
        computeStatus(hasNetwork, false)
      }
    } else {
      // User forces offline
      computeStatus(hasNetwork, true)
    }
  }

  return {
    hasNetwork, // true if browser reports network available
    isOnline, // effective online state (TanStack Query)
    isOffline: !isOnline, // effective offline state
    toggleOnline // call with `true|false` to switch modes
  }
}
