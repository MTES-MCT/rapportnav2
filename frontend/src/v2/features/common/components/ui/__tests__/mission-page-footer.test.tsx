import { vi } from 'vitest'
import { render } from '../../../../../../test-utils'
import { useOnlineManager } from '../../../hooks/use-online-manager.tsx'
import useMission from '../../../services/use-mission.tsx'
import MissionPageFooter from '../mission-page-footer'

vi.mock('../../../services/use-mission.tsx', () => ({
  default: vi.fn()
}))

vi.mock('../../../hooks/use-online-manager.tsx', () => ({
  useOnlineManager: vi.fn()
}))

const exitMission = vi.fn()

describe('MissionPageFooter', () => {
  beforeEach(() => {
    vi.clearAllMocks()
    vi.clearAllTimers()
    vi.mocked(useOnlineManager).mockReturnValue({ isOnline: true } as any)
    vi.mocked(useMission).mockReturnValue({ dataUpdatedAt: 0 } as any)
  })
  it('should match the snapshot', () => {
    const wrapper = render(<MissionPageFooter exitMission={exitMission} missionId={'missionId'} />)
    expect(wrapper).toMatchSnapshot()
  })
  describe('synchro text', () => {
    beforeEach(() => {
      vi.clearAllMocks()
      vi.clearAllTimers()
    })
    it('should say offline when offline', () => {
      vi.mocked(useOnlineManager).mockReturnValue({ isOnline: false } as any)
      const wrapper = render(<MissionPageFooter exitMission={exitMission} missionId={'missionId'} />)
      expect(wrapper.getByText('Connexion indisponible', { exact: false })).toBeInTheDocument()
    })
    it('should say online when online', () => {
      const wrapper = render(<MissionPageFooter exitMission={exitMission} missionId={'missionId'} />)
      expect(wrapper.getByText('Connexion disponible', { exact: false })).toBeInTheDocument()
    })
    it('should not show last sync text when dataUpdatedAt falsy', () => {
      vi.mocked(useMission).mockReturnValue({ dataUpdatedAt: 0 } as any)
      const wrapper = render(<MissionPageFooter exitMission={exitMission} missionId={'missionId'} />)
      expect(wrapper.queryAllByAltText('Dernière synchronisation ', { exact: false })).toHaveLength(0)
    })
    it('should show last sync text when dataUpdatedAt truthy', () => {
      vi.mocked(useMission).mockReturnValue({ dataUpdatedAt: 1234567890 } as any)
      const wrapper = render(<MissionPageFooter exitMission={exitMission} missionId={'missionId'} />)
      expect(wrapper.getByText('Dernière synchronisation ', { exact: false })).toBeInTheDocument()
    })
  })
})
