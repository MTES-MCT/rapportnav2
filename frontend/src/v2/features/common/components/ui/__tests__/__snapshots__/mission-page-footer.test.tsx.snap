// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MissionPageFooter > should match the snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-lehtBJ ezcbjO"
      >
        <div
          class="rs-stack"
          style="align-items: center; justify-content: space-between; flex-direction: row; height: 100%;"
        >
          <div
            class="rs-stack-item"
            style="padding-left: 1rem;"
          >
            <button
              class="sc-eknHtZ sc-jYczwO eIVVCb jxeylY Element-Button"
              disabled=""
              title="Cette fonctionnalité n'a pas encore été implémentée"
              type="button"
            >
              <span
                class="sc-hjsuWn eVTFbL Element-IconBox"
              >
                <svg
                  height="20"
                  viewBox="0 0 20 20"
                  width="20"
                >
                  <g
                    transform="translate(-23.798 -14.096)"
                  >
                    <g>
                      <rect
                        fill="currentColor"
                        height="6"
                        transform="translate(31.548 22.096)"
                        width="1.5"
                      />
                      <rect
                        fill="currentColor"
                        height="6"
                        transform="translate(34.548 22.096)"
                        width="1.5"
                      />
                      <path
                        d="M41.8,17.6h-4V16.1h-8v1.5h-4v2h1.5V31.1a1,1,0,0,0,1,1h11a1,1,0,0,0,1-1V19.6h1.5ZM38.3,30.1h-9V19.6h9Z"
                        fill="currentColor"
                      />
                    </g>
                    <rect
                      fill="none"
                      height="20"
                      transform="translate(23.798 14.096)"
                      width="20"
                    />
                  </g>
                </svg>
              </span>
              <span
                class="sc-pYNGo enBrtN"
              >
                Supprimer la mission
              </span>
            </button>
          </div>
          <div
            class="rs-stack-item"
          >
            <div
              class="rs-stack"
              style="align-items: center; flex-direction: row;"
            >
              <div
                class="rs-stack-item"
              >
                <p
                  class="sc-evkzZa sc-fFmWCW fjwyIl hLoVng"
                  font-size="11"
                >
                  Connexion disponible 
                   
                </p>
              </div>
            </div>
          </div>
          <div
            class="rs-stack-item"
            style="padding-left: 1rem;"
          >
            <div
              class="rs-stack"
              style="align-items: center; justify-content: center; flex-direction: row; gap: 0.5rem;"
            >
              <div
                class="rs-stack-item"
              >
                <p
                  class="sc-evkzZa sc-enPhjR bIoXEW esqQrj"
                  color="#CCCFD6"
                  font-size="13"
                  weight="normal"
                >
                  Hors ligne
                </p>
              </div>
              <div
                class="rs-stack-item"
              >
                <div
                  class="sc-fYmhhH jaaEau Element-Field Field-Toggle sc-kbeVdv hLmyn"
                >
                  <label
                    class="sc-gahYZc eeWxSe Element-Label"
                    for="online-toggle"
                  />
                  <label
                    class="sc-iyBeIh jlfYnm rs-toggle rs-toggle-checked rs-toggle-disabled"
                    title="Nous ne pouvez pas passer en ligne, connection non détectée."
                  >
                    <input
                      aria-checked="true"
                      aria-disabled="true"
                      aria-label="Ouvrir"
                      checked=""
                      class="rs-toggle-input"
                      disabled=""
                      name="online-toggle"
                      role="switch"
                      type="checkbox"
                    />
                    <span
                      class="rs-toggle-presentation"
                    >
                      <span
                        class="rs-toggle-inner"
                      >
                        <span
                          class="sc-hjsuWn fOmxtd Element-IconBox"
                        >
                          <svg
                            height="20"
                            viewBox="0 0 20 20"
                            width="20"
                          >
                            <g
                              transform="translate(40 -82)"
                            >
                              <g
                                fill="none"
                                stroke-miterlimit="10"
                              >
                                <path
                                  d="M-23.99,86.7l-7.778,7.778-4.243-4.243-1.414,1.414,4.243,4.243,1.414,1.414,9.192-9.192Z"
                                  stroke="none"
                                />
                                <path
                                  d="M -23.98955917358398 86.69669342041016 L -22.57537841796875 88.11093139648438 L -31.76775932312012 97.30330657958984 L -37.42462158203125 91.64644622802734 L -36.01044082641602 90.23220825195313 L -31.76775932312012 94.47487640380859 L -23.98955917358398 86.69669342041016 Z"
                                  fill="currentColor"
                                  stroke="none"
                                />
                              </g>
                              <rect
                                fill="none"
                                height="20"
                                transform="translate(-40 82)"
                                width="20"
                              />
                            </g>
                          </svg>
                        </span>
                      </span>
                    </span>
                  </label>
                </div>
              </div>
              <div
                class="rs-stack-item"
              >
                <p
                  class="sc-evkzZa sc-enPhjR kywkhf esqQrj"
                  color="#29B361"
                  font-size="13"
                  weight="bold"
                >
                  En ligne
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="rs-toast-provider"
    >
      <div
        class="rs-toast-container rs-toast-container-top-center"
      />
      <div
        class="rs-toast-container rs-toast-container-bottom-center"
      />
      <div
        class="rs-toast-container rs-toast-container-top-start"
      />
      <div
        class="rs-toast-container rs-toast-container-top-end"
      />
      <div
        class="rs-toast-container rs-toast-container-bottom-start"
      />
      <div
        class="rs-toast-container rs-toast-container-bottom-end"
      />
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-lehtBJ ezcbjO"
    >
      <div
        class="rs-stack"
        style="align-items: center; justify-content: space-between; flex-direction: row; height: 100%;"
      >
        <div
          class="rs-stack-item"
          style="padding-left: 1rem;"
        >
          <button
            class="sc-eknHtZ sc-jYczwO eIVVCb jxeylY Element-Button"
            disabled=""
            title="Cette fonctionnalité n'a pas encore été implémentée"
            type="button"
          >
            <span
              class="sc-hjsuWn eVTFbL Element-IconBox"
            >
              <svg
                height="20"
                viewBox="0 0 20 20"
                width="20"
              >
                <g
                  transform="translate(-23.798 -14.096)"
                >
                  <g>
                    <rect
                      fill="currentColor"
                      height="6"
                      transform="translate(31.548 22.096)"
                      width="1.5"
                    />
                    <rect
                      fill="currentColor"
                      height="6"
                      transform="translate(34.548 22.096)"
                      width="1.5"
                    />
                    <path
                      d="M41.8,17.6h-4V16.1h-8v1.5h-4v2h1.5V31.1a1,1,0,0,0,1,1h11a1,1,0,0,0,1-1V19.6h1.5ZM38.3,30.1h-9V19.6h9Z"
                      fill="currentColor"
                    />
                  </g>
                  <rect
                    fill="none"
                    height="20"
                    transform="translate(23.798 14.096)"
                    width="20"
                  />
                </g>
              </svg>
            </span>
            <span
              class="sc-pYNGo enBrtN"
            >
              Supprimer la mission
            </span>
          </button>
        </div>
        <div
          class="rs-stack-item"
        >
          <div
            class="rs-stack"
            style="align-items: center; flex-direction: row;"
          >
            <div
              class="rs-stack-item"
            >
              <p
                class="sc-evkzZa sc-fFmWCW fjwyIl hLoVng"
                font-size="11"
              >
                Connexion disponible 
                 
              </p>
            </div>
          </div>
        </div>
        <div
          class="rs-stack-item"
          style="padding-left: 1rem;"
        >
          <div
            class="rs-stack"
            style="align-items: center; justify-content: center; flex-direction: row; gap: 0.5rem;"
          >
            <div
              class="rs-stack-item"
            >
              <p
                class="sc-evkzZa sc-enPhjR bIoXEW esqQrj"
                color="#CCCFD6"
                font-size="13"
                weight="normal"
              >
                Hors ligne
              </p>
            </div>
            <div
              class="rs-stack-item"
            >
              <div
                class="sc-fYmhhH jaaEau Element-Field Field-Toggle sc-kbeVdv hLmyn"
              >
                <label
                  class="sc-gahYZc eeWxSe Element-Label"
                  for="online-toggle"
                />
                <label
                  class="sc-iyBeIh jlfYnm rs-toggle rs-toggle-checked rs-toggle-disabled"
                  title="Nous ne pouvez pas passer en ligne, connection non détectée."
                >
                  <input
                    aria-checked="true"
                    aria-disabled="true"
                    aria-label="Ouvrir"
                    checked=""
                    class="rs-toggle-input"
                    disabled=""
                    name="online-toggle"
                    role="switch"
                    type="checkbox"
                  />
                  <span
                    class="rs-toggle-presentation"
                  >
                    <span
                      class="rs-toggle-inner"
                    >
                      <span
                        class="sc-hjsuWn fOmxtd Element-IconBox"
                      >
                        <svg
                          height="20"
                          viewBox="0 0 20 20"
                          width="20"
                        >
                          <g
                            transform="translate(40 -82)"
                          >
                            <g
                              fill="none"
                              stroke-miterlimit="10"
                            >
                              <path
                                d="M-23.99,86.7l-7.778,7.778-4.243-4.243-1.414,1.414,4.243,4.243,1.414,1.414,9.192-9.192Z"
                                stroke="none"
                              />
                              <path
                                d="M -23.98955917358398 86.69669342041016 L -22.57537841796875 88.11093139648438 L -31.76775932312012 97.30330657958984 L -37.42462158203125 91.64644622802734 L -36.01044082641602 90.23220825195313 L -31.76775932312012 94.47487640380859 L -23.98955917358398 86.69669342041016 Z"
                                fill="currentColor"
                                stroke="none"
                              />
                            </g>
                            <rect
                              fill="none"
                              height="20"
                              transform="translate(-40 82)"
                              width="20"
                            />
                          </g>
                        </svg>
                      </span>
                    </span>
                  </span>
                </label>
              </div>
            </div>
            <div
              class="rs-stack-item"
            >
              <p
                class="sc-evkzZa sc-enPhjR kywkhf esqQrj"
                color="#29B361"
                font-size="13"
                weight="bold"
              >
                En ligne
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
